modules/search/search_results.go

```diff
@@ -0,0 +1,101 @@
+package search
+
+import (
+	"context"
+
+	"github.com/grafana/tempo/pkg/tempopb"
+	"go.uber.org/atomic"
+)
+
+// SearchResults eases performing a highly parallel search by funneling all results into a single
+// channel that is easy to consume, signaling workers to quit early as needed, and collecting
+// metrics.
+type SearchResults struct {
+	resultsCh   chan *tempopb.TraceSearchMetadata
+	doneCh      chan struct{}
+	quit        atomic.Bool
+	workerCount atomic.Int32
+
+	tracesInspected atomic.Uint32
+	bytesInspected  atomic.Uint64
+}
+
+func NewSearchResults() *SearchResults {
+	return &SearchResults{
+		resultsCh: make(chan *tempopb.TraceSearchMetadata),
+		doneCh:    make(chan struct{}),
+	}
+}
+
+// AddResult sends a search result from a search task (goroutine) to the receiver of the
+// the search results, i.e. the initiator of the search.  This function blocks until there
+// is buffer space in the results channel or if the task should stop searching because the
+// receiver went away or the given context is done. In this case true is returned.
+func (sr *SearchResults) AddResult(ctx context.Context, r *tempopb.TraceSearchMetadata) (quit bool) {
+	select {
+	case sr.resultsCh <- r:
+		return false
+	case <-ctx.Done():
+		return true
+	case <-sr.doneCh:
+		// This returns immediately once the done channel is closed.
+		return true
+	}
+}
+
+// Quit returns if search tasks should quit early. This can occur due to max results
+// already found, or other errors such as timeout, etc.
+func (sr *SearchResults) Quit() bool {
+	return sr.quit.Load()
+}
+
+// Results returns the results channel. Channel is closed when the search is complete.
+// Can be iterated by range like:
+//   for res := range sr.Results()
+func (sr *SearchResults) Results() <-chan *tempopb.TraceSearchMetadata {
+	return sr.resultsCh
+}
+
+// Close signals to all workers to quit, when max results is received and no more work is needed.
+// Called by the initiator of the search in a defer statement like:
+//     sr := NewSearchResults()
+//     defer sr.Close()
+func (sr *SearchResults) Close() {
+	// Closing done channel makes all subsequent and blocked calls to AddResult return
+	// quit immediately.
+	close(sr.doneCh)
+	sr.quit.Store(true)
+}
+
+// SetWorkerCount sets the number of workers that will be started
+func (sr *SearchResults) SetWorkerCount(count int32) {
+	sr.workerCount.Add(count)
+}
```

Comment:
@mdisibio fyi changed this because of -

```
tempo_1                     | panic: close of closed channel
tempo_1                     |
tempo_1                     | goroutine 1685 [running]:
tempo_1                     | github.com/grafana/tempo/modules/search.(*SearchResults).FinishWorker(0xc0010ec930)
tempo_1                     | 	/Users/annanay/Desktop/git/go/src/github.com/grafana/tempo/modules/search/search_results.go:83 +0x48
tempo_1                     | github.com/grafana/tempo/modules/ingester.(*instance).searchWAL.func1(0xc0010ec930, 0x29a06a8, 0xc0010ec8d0, 0xc000d18970, 0x1, 0x1, 0x295c0e0, 0xc000d1a720)
tempo_1                     | 	/Users/annanay/Desktop/git/go/src/github.com/grafana/tempo/modules/ingester/instance_search.go:98 +0xb9
tempo_1                     | created by github.com/grafana/tempo/modules/ingester.(*instance).searchWAL
tempo_1                     | 	/Users/annanay/Desktop/git/go/src/github.com/grafana/tempo/modules/ingester/instance_search.go:95 +0x165
```

Can move this logic around more if needed.